### PR TITLE
feat(sparse): native supernodal LDLᵀ with mixed-precision accumulator

### PR DESCRIPTION
**Pull request by Ravenwater**
## Summary

Adds a native, header-only **supernodal LDLᵀ** factorization for symmetric matrices: `A = PᵀL D Lᵀ P`. Columns sharing a nonzero structure are grouped into **dense panels (supernodes)**, so the bulk of the arithmetic becomes a dense block update of one panel by another — the natural granularity for mixed precision and the first native supernodal factorization in MTL5.

This is milestone 1 of the supernodal-for-mixed-precision effort (the safe, symmetric, no-pivoting target before any unsymmetric SuperLU-style work).

## Why LDLᵀ first

- Square-root free → friendlier to low precision and custom number types (posit/LNS) where `sqrt` is costly/less accurate; `D` is a natural place to study precision.
- Symmetric ⇒ no pivoting ⇒ supernodes follow directly from the elimination tree.

## The mixed-precision boundary

The dense panel is touched **only** through `mtl::math::accumulator_traits` (`add_product` / `value` / `clear`). So:
- The factor is **stored low** (`Value`) while every accumulation is carried in a higher-precision **`Accumulator`**.
- Each entry is rounded to the storage type **exactly once** when finalized (single-rounding) — works with a wider IEEE type *or* an add-only super-accumulator (Universal `quire`).
- Default `Accumulator == Value` is byte-identical to a plain factorization.

The numeric kernel emits a **standard CSC `L` + diagonal `D`**, so `solve` and the generic `iterative_refine` loop are reused unchanged. `supernodal_ldlt_solve_refined<float, double>(...)` factors low and recovers accuracy with a high-precision residual.

## What's included

| File | Purpose |
|------|---------|
| `include/mtl/sparse/analysis/supernodes.hpp` | ereach column patterns + fundamental supernode detection |
| `include/mtl/sparse/factorization/supernodal_ldlt.hpp` | symbolic, left-looking dense-panel numeric, solve, mixed-precision IR |
| `tests/unit/sparse/test_supernodal_ldlt.cpp` | symbolic, numeric vs scalar oracle, mixed precision, IR, edge cases |
| `docs/algorithms/supernodal-ldlt.md` | docs page (+ FILE_MAP entry) |

Reuses the existing symbolic frontend (elimination tree, postorder, AMD), `triangular_solve` kernels, `accumulator_traits`, and `iterative_refine`.

## Testing

- `ctest -R supernodal` — 8 cases / 66 assertions, all pass.
- Full suite: **126/126 pass**; new headers warning-clean under GCC and Clang `-Wall -Wextra`.
- Coverage: diagonal ⇒ all-singleton supernodes; dense ⇒ one panel; numeric agreement with scalar `sparse_ldlt` on tridiagonal / 2-D Laplacian / dense SPD (residual < 1e-12); wider accumulator never less accurate; float-factor + double-residual IR recovers near-double accuracy and beats float-only; 1×1 / diagonal / zero-pivot edge cases.

## Scope / follow-ups

- Only **fundamental** supernodes (the `relax` knob is accepted but reserved for relaxed amalgamation later).
- Solve reuses CSC kernels; a **blocked supernodal solve** is a future optimization.
- Unsymmetric supernodal LU (SuperLU-style, pivoting × supernodes) is a separate, later effort.

> Note: the approved plan suggested splitting numeric-only (PR1) from mixed-precision + IR (PR2). Since the mixed-precision boundary is the whole point and it's fully implemented and tested, this PR delivers both together as one cohesive header rather than landing code that would be immediately rewritten.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Added a native supernodal **LDLᵀ** factorization path for symmetric CSC matrices with one-shot solve and reusable symbolic/numeric workflows.
  * Support for mixed-precision factorization/accumulation plus mixed-precision iterative refinement for higher accuracy.
* **Documentation**
  * Added a new documentation page covering the factorization approach, API usage, validation, and stated scope/limitations.
* **Tests**
  * Added unit tests for symbolic supernode detection, numeric correctness, solve reuse, edge cases (including zero pivots), and mixed-precision + iterative refinement accuracy.

### CODE REVIEW

File: include/mtl/sparse/factorization/supernodal_ldlt.hpp [hunk 1]
```diff
@@ -0,0 +1,375 @@
...
+        panel_off[J].assign((m - w) * w, Value{0});
+        for (std::size_t c = 0; c < w; ++c) {
+            const Value dcc = AT::template value<Value>(P[c * m + c]);
+            if (dcc == Value{0}) {
+                for (std::size_t t = 0; t < m; ++t) rel[rows[t]] = NPOS;
+                throw std::runtime_error(
+                    "supernodal_ldlt_numeric: zero pivot at column "
+                    + std::to_string(f + c));
+            }
+            D[f + c] = dcc;
```
[the remark this answers is not in the document]
Comment (Ravenwater) [reply include/mtl/sparse/factorization/supernodal_ldlt.hpp:284]: Addressed in add222f. Broadened the guard to reject non-finite pivots too: `if (dcc == Value{0} || dcc - dcc != Value{0})`. The `dcc - dcc != 0` form is true for NaN and Inf and false for any finite value, so it stays dependency-free for custom number types (posit/LNS) — and catches Inf, which a plain `dcc != dcc` (NaN-only) would miss. I deliberately did **not** add a near-zero tolerance: LDLᵀ does no pivoting and supports indefinite systems, where a legitimately small pivot can occur, so rejecting small pivots would be incorrect.

File: tests/unit/sparse/test_supernodal_ldlt.cpp [hunk 2]
```diff
@@ -0,0 +1,280 @@
...
+    fac_f.solve(xf, b);
+    fac_d.solve(xd, b);
+
+    double res_f = rel_residual(A, xf, b);
+    double res_d = rel_residual(A, xd, b);
+
+    REQUIRE(std::isfinite(res_f));
+    REQUIRE(res_d < 1e-10);                   // double accumulation -> accurate factor
+    REQUIRE(res_d < res_f);                   // wider accumulator decisively better
+}
```
[the remark this answers is not in the document]
Comment (Ravenwater) [reply tests/unit/sparse/test_supernodal_ldlt.cpp:248]: Addressed in add222f. Strengthened to lock in a decisive (>=3-order) separation with floor assertions rather than a ratio: `REQUIRE(res_d < 1e-12); REQUIRE(res_f > 1e-9);`. I avoided the `res_f / res_d > 1e4` ratio because `res_d` is ~0 here (near-exact double-accumulated factor), which would make the division ill-conditioned/inf; the two floors enforce the same intent robustly.
